audio/src/passthrough_decoder.rs

```diff
@@ -0,0 +1,233 @@
+// Passthrough decoder for librespot
+use super::{AudioDecoder, AudioError, AudioPacket};
+use ogg::{OggReadError, Packet, PacketReader, PacketWriteEndInfo, PacketWriter};
+use std::fmt;
+use std::io::{Error, Read, Seek};
+
+fn read_headers<T: Read + Seek>(
+    rdr: &mut PacketReader<T>,
+) -> Result<(Vec<u8>, u32), PassthroughError> {
+    let mut header = PacketWriter::new(Vec::new());
+    let mut stream_serial = 0;
+
+    // search for ident, comment, setup
+    add_header(
+        1,
+        &mut header,
+        rdr,
+        &mut stream_serial,
+        PacketWriteEndInfo::EndPage,
+    )?;
+    add_header(
+        3,
+        &mut header,
+        rdr,
+        &mut stream_serial,
+        PacketWriteEndInfo::NormalPacket,
+    )?;
+    add_header(
+        5,
+        &mut header,
+        rdr,
+        &mut stream_serial,
+        PacketWriteEndInfo::EndPage,
+    )?;
+
+    // remove un-needed packets
+    rdr.delete_unread_packets();
+    return Ok((header.into_inner(), stream_serial));
+}
+
+fn add_header<T>(
+    code: u8,
+    header: &mut PacketWriter<Vec<u8>>,
+    rdr: &mut PacketReader<T>,
+    stream_serial: &mut u32,
+    info: PacketWriteEndInfo,
+) -> Result<(), PassthroughError>
+where
+    T: Read + Seek,
+{
+    let mut pck: Packet = rdr.read_packet_expected()?;
+    while *stream_serial != 0 && pck.stream_serial() != *stream_serial {
+        pck = rdr.read_packet_expected()?;
+    }
+
+    *stream_serial = pck.stream_serial();
+
+    let pkt_type = pck.data[0];
+    debug!("Vorbis header type{}", &pkt_type);
+
+    // all headers are mandatory
+    if pkt_type != code {
+        return Err(PassthroughError::NotVorbisError);
+    }
+
+    let absgp_page = pck.absgp_page();
+    header.write_packet(
+        pck.data.into_boxed_slice(),
+        *stream_serial,
+        info,
+        absgp_page,
+    )?;
+
+    return Ok(());
+}
+
+static mut BASEGP_PAGE: u64 = 0;
```

Comment:
It is a static because it survives decoder instances and it is stitching together the global timeline. If not, multiple tracks will have an incorrect position and the resulting file will not be readable

Reply:
Ah right. So is the idea that for the lifetime of the process the granule only ever gets incremented?

Reply:
I don't understand what apsgp does, so I might be wrong. But librespot could be used as library, and multiple PassthroughDecoders could be created at the same time for whatever reason. What would happen then? There isn't any protection against concurrent access to `BASEGP_PAGE`, is it?

Reply:
> Ah right. So is the idea that for the lifetime of the process the granule only ever gets bigger?
>
Yes

Reply:
> I don't understand what apsgp does, so I might be wrong. But librespot could be used as library, and multiple PassthroughDecoders could be created at the same time for whatever reason. What would happen then? There isn't any protection against concurrent access to `BASEGP_PAGE`, is it?
>
There is probably no perfect answer. This relates to granule position's page. It is used to seek into file as well as a way to know the duration of an ogg file:  the granule of the last page of the file give you the duration, there is no other way. So if I don't have that incremented properly, then duration is wrong and players can be totally confused. There is no perfect solution anyway b/c when a seek is performed, the granule will overshoot, but that's better than undershoot that lead to premature ending. Re having that as a global, I agree and I don't like globals, but otherwise I would need to offer a "context" which is much more complications and non-compatibilty with existing code. I could have created another new() with an extra parameter to reset it but I think Rust does not offer overloading/mangling

Reply:
Ah, it's clear for me now.

But using `unsafe` in a library such that undefined behaviour could occur if you use it wrong should definitely be avoided.

You could use 
```rust

static BASEGP_PAGE: AtomicU64::new(0);

// Adding
BASEGP_PAGE.fetch_add(n, Ordering::Relaxed);
```

Even though it doesn't make sense to use multiple instance at the same time, it's at least no UB.

Reply:
done